Added spec for 'package updater' tools (#12563)
Azure/azure-sdk-tools: tools/azsdk-cli/docs/specs/6-package-updater.spec.md

- Error-prone when performed manually (missed version increments, stale README examples, inconsistent metadata presented across different files, like `pom.xml`, `_metadata.json`).
- A barrier for integrating an AI agent workflow because required parameters and step order are not encapsulated as composable tools.

### Why This Matters

Without a unified tool layer:

- Contributors spend time re-learning differing per-language conventions.
- CI or release pipelines fail late due to missing metadata or mismatched versions.
- Agents cannot reliably assist because state transitions ("now bump version", "now update changelog") aren't formalized.
- Management-plane vs data-plane discrepancies introduce confusion (e.g., when to bump versions or generate changelog entries).

Formalizing Stage 6 as cohesive tools improves reliability, supports automation-first workflows, and prepares for future LLM-assisted tasks.

---

## Goals and Exceptions/Limitations

### Goals

- [ ] Provide three composable, idempotent tools exposed via CLI and MCP agent: Changelog Update, Version Update, Metadata Update.
- [ ] Normalize execution order and expected outputs across all supported languages.

> - [ ] Encode management-plane vs data-plane behavioral differences explicitly (especially version & changelog logic).

Reviewer: great that this will help further standardize on mgmt and data-plane differences and make them more known/explicit.

- [ ] Emit structured JSON results (status + message + next steps hints) enabling orchestration by AI agents and pipelines.
- [ ] Support configuration-driven per-language scripting (path or inline) without recompiling the tool.

### Exceptions

| Exception | Description | Impact | Workaround |
|---|---|---|---|
| Data‑Plane Deferred Version Bump | Data‑plane packages typically defer version adjustment and release‑date updates until the release‑preparedness phase. | The `update-version` tool may be a no‑op for data‑plane flows during Stage 6. | Keep version changes manually for milestone 1. |
| Changelog Auto‑Generation (Data‑Plane) | Automatic changelog generation is not performed for data‑plane packages; manual editing is expected. | The `update-changelog-content` tool should not auto‑generate entries for data‑plane packages and should instead surface guidance. | Return a `noop` result with `next_steps` prompting manual edits. Consider future LLM integration to automate this post milestone 1. |

### Language-Specific Limitations

| Language   | Limitation | Impact | Workaround |
|------------|------------|--------|------------|

> | JavaScript | Code formatting isn't hooked up to the emitter | Noisy diffs or linter failures | Run code formatting step in the `update-metadata` tool |

Author: The formatting step will be moved to the JS emitter based on the discussion with @MaryGao.

> | JavaScript | Code formatting isn't hooked up to the emitter | Noisy diffs or linter failures | Run code formatting step in the `update-metadata` tool |

Author: Thanks @MaryGao for confirming. I removed it from the spec.

---

## Design Proposal

### Overview

Implement three individual CLI + MCP tools. Each tool:

1. Resolves package path & language.
2. Loads configuration for the language repository.
3. If a script is configured, executes the script.
4. Otherwise falls back the CLI's built-in implementation for that language.
5. Aggregates structured result JSON with a `result`, human-readable `message`, and optional `next_steps` hint guiding subsequent tool invocation.

> Note: see the [repo-tooling-contract](./99-repo-tooling-contract.md) spec for more details.

Reviewer 2: Let's not reference this spec yet until we get it finalized.

Instead based on your current implementation at https://github.com/Azure/azure-sdk-tools/pull/12761 you are using the existing swagger_to_sdk_config.json configuration so please add that here with an example. That will also help answer some of the questions in your implementation PR.

Author: Removed this link and added the configuration section and a sample.

| Changelog Auto‑Generation (Data‑Plane) | Automatic changelog generation is not performed for data‑plane packages; manual editing is expected. | The `update-changelog` tool should not auto‑generate entries for data‑plane packages and should instead surface guidance. | Return a `noop` result with `next_steps` prompting manual edits. Consider future LLM integration to automate this post milestone 1. |

### Language-Specific Limitations

| Language   | Limitation | Impact | Workaround |
|------------|------------|--------|------------|
| JavaScript | Code formatting isn't hooked up to the emitter | Noisy diffs or linter failures | Run code formatting step in the `update-metadata` tool |

---

## Design Proposal

### Overview

Implement four individual CLI + MCP tools. Each tool:

1. Resolves package path & language.
2. Loads configuration for the language repository.
3. If a script is configured, executes the script.
4. Otherwise falls back the CLI's built-in implementation for that language.
5. Aggregates structured result JSON with a `result`, human-readable `message`, and optional `next_steps` hint guiding subsequent tool invocation.

> Provides a singular CLI + MCP command that orchestrates the four tools (update-ci → update-changelog → update-version → update-metadata) and returns a combined JSON summary (post milestone 1).

Author: Removed it from the spec since it's not part of the milestone1.

{"result":"succeeded", "message":"Changelog entry added for 2025-10-16", "next_steps": "update version"}
```

```json
{"result":"noop", "message":"Data-plane changelog untouched", "next_steps": "Manually edit the changelog, and update version"}
```

Failure Modes:

- Missing `CHANGELOG.md` → create template (mgmt-plane) or noop (data-plane).
Note: we should consider adding recovery support for the generation scenario after milestone 1.
- Script failure → `failed` with error output.

#### 2. Version Update Tool

Name (CLI): `azsdk package update-version`

Name (MCP): `azsdk_package_update_version`

Purpose: Increment and propagate package version across all authoritative version surfaces (e.g., `pom.xml`, `_version.py`, `*.csproj`, `package.json`, `go.mod`, `CHANGELOG.md`).

Inputs:

- `--package-path <abs path>` (required)
- `--release-type <stable|beta>` (optional)
- `--version` (optional)
- `--release-date` (optional)

Execution Steps:

- for mgmt-plane SDKs:

 1. Compute a version based on the given information.
 2. Apply updates across language-specific files.
 3. Validate consistency (all files now share identical new version).

- for data-plane SDKs:

 1. return `noop`.

>  2. use current version from the metadata file or from user input (support may be post-milestone-1).

Reviewer 3: So the tool returns a noop and then you run it again?

- [ ] Metadata files are updated accordingly.

    | Language | Exception |
    |---|---|
    |Java|No additional files need updating|
    |JavaScript|No additional files need updating|
    |Go|No additional files need updating|

- [ ] Cross-language scripts execute without manual intervention.
- [ ] Errors surface actionable mitigation text.

---

## Agent Prompts

### Scenario 1: Update Management-Plane Package (existing SDK) for Release

**Prompt:**

```text
Update the package at /home/dev/sdk/healthdataaiservices/Azure.ResourceManager.HealthDataAIServices for a stable release.
```

**Expected Agent Activity:**

1. Run changelog update.
2. Run version update with inferred `--release-type stable`.
3. Run metadata update (README, API export, formatting if needed) and summarize JSON results.

### Scenario 2: Update Version for Stable Release (Data-Plane)

**Prompt:**

```text
Update the version and release date for package at /repo/sdk/healthdataaiservices/Azure.Health.Deidentification.
```

**Expected Agent Activity:**

1. Confirm with user on the version to use.
2. Run version update with the confirmed version.
3. Report the result.

> ### Scenario 3: Create CI Config for New SDK

Reviewer 3: I dont see this CI update tool outlined above - is this something already in the azsdkcli?

Author: No, they were in the spec before. I removed them from the spec based on the feedback since it won't be included in milestone 1.

- [ ] Emit structured JSON results (status + message + next steps hints) enabling orchestration by AI agents and pipelines.
- [ ] Support configuration-driven per-language scripting (path or inline) without recompiling the tool.

### Exceptions

| Exception | Description | Impact | Workaround |
|---|---|---|---|
| Data‑Plane Deferred Version Bump | Data‑plane packages typically defer version adjustment and release‑date updates until the release‑preparedness phase. | The `update-version` tool may be a no‑op for data‑plane flows during Stage 6. | Keep version changes manually for milestone 1. |
| Changelog Auto‑Generation (Data‑Plane) | Automatic changelog generation is not performed for data‑plane packages; manual editing is expected. | The `update-changelog` tool should not auto‑generate entries for data‑plane packages and should instead surface guidance. | Return a `noop` result with `next_steps` prompting manual edits. Consider future LLM integration to automate this post milestone 1. |

### Language-Specific Limitations

| Language   | Limitation | Impact | Workaround |
|------------|------------|--------|------------|

> | JavaScript | Code formatting isn't hooked up to the emitter | Noisy diffs or linter failures | Run code formatting step in the `update-metadata.ps1` script |

Reviewer 3: in light of our recent conversations I would argue that none of these should then be scripts

- Python: `_metadata.json` and `pyproject.toml`.
- JavaScript: N/A (no targets).
- Go: N/A (no targets).

Inputs:

- `--package-path <abs path>` (required)

Execution Steps:

1. Apply updates to authoritative metadata files (`_metadata.json`, `pom.xml`, `package.json`, `go.mod`). Only overwrite a file when the new content differs from the existing content. If an expected file is missing, create a minimal safe skeleton when possible and warn otherwise.

2. Return a structured summary in the tool JSON result including `changed_files` (array of relative paths) and `changed_files_count` (integer). When no files were modified return `changed_files: []` and `changed_files_count: 0` and set `message` to indicate "no changes" (`changed_files` may be supported post-milestone-1).

Outputs:

```json
{"result":"succeeded", "changed_files":["..\\pom.xml","C:\\dev\\repos\\azure-sdk-for-java\\pom.xml"], "changed_files_count": 2, "message":"All relevant pom.xml files have been updated."}
```

```json
{"result":"failed","message":"API export failed: missing dotnet tool", "next_steps": "call verify-setup tool"}
```

Failure Modes:

- Missing expected file (e.g., `_metadata.json`) → report error.

> ##### Update Metadata script

Reviewer 2: ```suggestion
##### Update Metadata tool
```

Lets not imply it is a script necessarily.

- CI or release pipelines fail late due to missing metadata or mismatched versions.
- Agents cannot reliably assist because state transitions ("now bump version", "now update changelog") aren't formalized.
- Management-plane vs data-plane discrepancies introduce confusion (e.g., when to bump versions or generate changelog entries).

Formalizing Stage 6 as cohesive tools improves reliability, supports automation-first workflows, and prepares for future LLM-assisted tasks.

---

## Goals and Exceptions/Limitations

### Goals

- [ ] Provide four composable, idempotent tools exposed via CLI and MCP agent: CI Update, Changelog Update, Version Update, Metadata Update.
- [ ] Normalize execution order and expected outputs across all supported languages.
- [ ] Encode management-plane vs data-plane behavioral differences explicitly (especially version & changelog logic).
- [ ] Emit structured JSON results (status + message + next steps hints) enabling orchestration by AI agents and pipelines.
- [ ] Support configuration-driven per-language scripting (path or inline) without recompiling the tool.

### Exceptions

| Exception | Description | Impact | Workaround |
|---|---|---|---|

> | Data‑Plane Deferred Version Bump | Data‑plane packages typically defer version adjustment and release‑date updates until the release‑preparedness phase. | The `update-version` tool may be a no‑op for data‑plane flows during Stage 6. | Keep version changes manually for milestone 1. |

Reviewer 4: Do we have any ideas for how we will handle this after scenario 1? 

My thought is that the data-plane logic will be put in the same tooling, and it'll be up to our instructions to let the agent know when this tool should be called for data-plane vs mgmt-plane.

Author: The logic to determine the SDK type is within the tool code since the tool is expected to run in multiple scenarios: agent, CI pipeline, and local CLI.

| Changelog Auto‑Generation (Data‑Plane) | Automatic changelog generation is not performed for data‑plane packages; manual editing is expected. | The `update-changelog` tool should not auto‑generate entries for data‑plane packages and should instead surface guidance. | Return a `noop` result with `next_steps` prompting manual edits. Consider future LLM integration to automate this post milestone 1. |

### Language-Specific Limitations

| Language   | Limitation | Impact | Workaround |
|------------|------------|--------|------------|

> | JavaScript | Code formatting isn't hooked up to the emitter | Noisy diffs or linter failures | Run code formatting step in the `update-metadata.ps1` script |

Reviewer 4: Are there any issues with waiting till the Package Metadata & Docs Updates stage to run this code formatting? 
Will needed formatting changes force users to have to go back to the TypeSpec/Code customizations, testing, samples, or validation phase?

**Workaround:** Initial version of milestone 1 remains user decided and manual edits.

#### Exception 2: Changelog Auto-Generation (Data-Plane)

**Description:** No automatic generation for data-plane; manual editing expected.
**Impact:** Tool returns success with a `next_steps` message prompting manual edits.
**Workaround:** Future LLM integration (post milestone 1).

#### Language-Specific Limitations

| Language   | Limitation | Impact | Workaround |
|------------|------------|--------|------------|
| JavaScript | Code formatting isn't hooked up to the emitter | Noisy diffs or linter failures | Run code formatting step in the `update-metadata.ps1` script |

---

## Design Proposal

### Overview

Implement four individual CLI + MCP tools. Each tool:

1. Resolves package path & language.
2. Loads the configuration for the language.
3. Chooses inline command over script path if both present.
4. Executes the relevant sub-script(s) in an ordered, fault-isolated manner.
5. Aggregates structured result JSON with a `result`, human-readable `message`, and optional `next_steps` hint guiding subsequent tool invocation.

> Provide an aggregated CLI + MCP command that deterministically orchestrates the four tools (update-ci → update-changelog → update-version → update-metadata) and returns a combined JSON summary.

Reviewer 3: ```suggestion
Provides a singular CLI + MCP command that orchestrates the four tools (update-ci → update-changelog → update-version → update-metadata) and returns a combined JSON summary.
```
nit:

| Parameter | Description | Type | Required |
|-----------|-------------|------:|:--------:|
| `sdkRepoPath` | Absolute path to the root folder of the local SDK repository. | string | Yes |
| `packagePath` | Absolute path to the root folder of the local SDK project (package). | string | Yes |
| `sdkReleaseType` | Specifies whether the next version is `beta` or `stable`. | string | No |
| `sdkVersion` | Specifies the next version to set (explicit version string). | string | No |
| `sdkReleaseDate` | Release date to write into changelog (YYYY-MM-DD). | string | No |

Note: For management-plane (mgmt) packages, at least one of `sdkVersion` or `sdkReleaseType` MUST be provided; the script will fail with an error if neither is supplied or validation fails.

**Example invocations:**

```powershell
update-version.ps1 \
      -sdkRepoPath C:\dev\repos\azure-sdk-for-net \
      -packagePath C:\dev\repos\azure-sdk-for-net\services\newservice \
      -sdkReleaseType stable \
      -sdkReleaseDate 2025-10-17
```

#### 3. Metadata Update Tool

Name (CLI): `azsdk package metadata update`

Name (MCP): `azsdk_package_metadata_update`

> Purpose: Update additional language-specific metadata and documentation, such as `_metadata.json`, `pom.xml` etc., export API surface (.NET).

Reviewer 5: Python left a comment and has been resolved.

> Purpose: Update additional language-specific metadata and documentation, such as `_metadata.json`, `pom.xml` etc., export API surface (.NET).

Reviewer 6: For .NET Management plain: api view files under $packagePath\api

Inputs:

- `--package-path <abs path>` (required)

Behavior Steps:

> 1. Apply updates to authoritative metadata files (`_metadata.json`, `pom.xml`, `package.json`, `go.mod`). Only overwrite a file when the new content differs from the existing content. If an expected file is missing, create a minimal safe skeleton when possible and warn otherwise.

Author: We should warn at minimum. For some files like go.mod, the script may be able to re-create the file. If the files are generated by emitter, they are not expected to be re-created here.

> 1. Apply updates to authoritative metadata files (`_metadata.json`, `pom.xml`, `package.json`, `go.mod`). Only overwrite a file when the new content differs from the existing content. If an expected file is missing, create a minimal safe skeleton when possible and warn otherwise.

Reviewer 7: @tadelesh  I didn't follow. Are you saying that for Go in management plane the versions will be updated by each specific tool?

> 2. Return a structured summary in the tool JSON result including `changed_files` (array of relative paths) and `changed_count` (integer). When no files were modified return `changed_files: []` and `changed_count: 0` and set `message` to indicate "no changes (idempotent)" (`changed_files` may be supported post-milestone-1).

Reviewer 3: ```suggestion
2. Return a structured summary in the tool JSON result including `changed_files` (array of relative paths) and `changed_count` (integer). When no files were modified return `changed_files: []` and `changed_count: 0` and set `message` to indicate "no changes" (`changed_files` may be supported post-milestone-1).
```
nit not sure the value that added

> 2. Return a structured summary in the tool JSON result including `changed_files` (array of relative paths) and `changed_count` (integer). When no files were modified return `changed_files: []` and `changed_count: 0` and set `message` to indicate "no changes (idempotent)" (`changed_files` may be supported post-milestone-1).

Author: Updated to `changes_files_count`. It's a little redundant but gives clients, pipelines and telemetry an immediate, cheap summary without parsing the array.

```

```json
{"result":"failed","message":"API export failed: missing dotnet tool; next_steps: call verify-setup tool"}
```

Failure Modes:

- Missing expected file (e.g., `_metadata.json`) → report error.

##### Update Metadata script

When the tool executes the language-specific `update-metadata` script it will pass the following parameters. All paths are absolute.

| Parameter | Description | Type | Required |
|-----------|-------------|-----:|:--------:|
| `sdkRepoPath` | Absolute path to the root folder of the local SDK repository. | string | Yes |
| `packagePath` | Absolute path to the root folder of the local SDK project (package). | string | Yes |

**Example invocations:**

```powershell
update-metadata.ps1 \
      -sdkRepoPath C:\dev\repos\azure-sdk-for-net \
      -packagePath C:\dev\repos\azure-sdk-for-net\services\newservice
```

#### 4. CI Update Tool

Name (CLI): `azsdk package ci update`

Name (MCP): `azsdk_package_ci_update`

> **Note:** Support for this tool is planned for post‑milestone-1.

Purpose: Create or update CI pipeline configuration.

Inputs:

- `--package-path <abs path>` (required) – Absolute path to the root of the SDK package.

> Behavior:

Reviewer 3: nit 
```suggestion
Behavior Steps:
```

Behavior:

- Validate the provided `packagePath` and fail fast if invalid.
- Run `update-ci` first to ensure CI config is present/updated.
- Run `update-changelog` next; if it returns `failed`, stop and return the error.
- Run `update-version` after changelog succeeds; support `noop` for data-plane flows and propagate `failed` results upward.
- Run `update-metadata` last to apply remaining metadata and documentation updates.
- Aggregate each tool's JSON result into a top-level JSON summary with per-tool results, overall `result` (succeeded/failed), and an ordered `next_steps` hint.

Failure semantics:

- If any tool returns `failed`, the aggregated tool should stop execution and return `failed` with the failing tool's details.

Example aggregated output (successful):

```json
{
      "result": "succeeded",
      "steps": [
            {"tool":"update-ci", "result":"succeeded", "ci-file": "../ci.yml"},
            {"tool":"update-changelog", "result":"succeeded"},
            {"tool":"update-version", "result":"succeeded", "version":"1.2.0", "release-date": "2025-10-17"},
            {"tool":"update-metadata", "result":"succeeded"}
      ],

>       "message": "Package updates complete; next_steps: open PR"

Reviewer 3: nit I believe next-steps would be a separate entry 
```suggestion
      "message": "Package updates complete",
      "next_steps": "open PR"
```

- [ ] Errors surface actionable mitigation text.

---

## Agent Prompts

### Scenario 1: Update Management-Plane Package (existing SDK) for Release

**Prompt:**

```text
Update the package at /home/dev/sdk/healthdataaiservices/Azure.ResourceManager.HealthDataAIServices for a stable release.
```

**Expected Agent Activity:**

1. Run changelog update (auto-generation script).
2. Run version update with inferred `--release-type stable`.
3. Run metadata update (README, API export, formatting if needed) and summarize JSON results.

### Scenario 2: Update Version for Stable Release (Data-Plane)

**Prompt:**

```text
Update the version for package at /repo/sdk/healthdataaiservices/Azure.Health.Deidentification.
```

**Expected Agent Activity:**

1. Confirm with user on the version to use.
2. Run version update with the confirmed version.
3. Report the result.

### Scenario 3: Create CI Config for New SDK

**Prompt:**

```text
Add initial CI configuration for a new SDK at /work/sdk/new-service/new-package.
```

**Expected Agent Activity:**

1. Run CI update tool.
2. Report the created configuration file.
3. Suggest running update-changelog tool next.

---

> ## CLI Commands

Reviewer 3: nit: next-steps is a property in the response not a part of the message https://github.com/Azure/azure-sdk-tools/blob/06abfe4918dafe5ba5c6b4f7251b33589992ae65/tools/azsdk-cli/Azure.Sdk.Tools.Cli/Models/Responses/CommandResponse.cs#L43

# Spec: 6 Package - package-update

> ## Table of Contents

Author: I'd like to simplify the spec name to `package update`, since it aligns better with the aggregated tool name `package update-all`. But I'm open to use the original name. @ronniegeraghty what do you think?

> ## Table of Contents

Reviewer 4: Right now, I'd say align with the tool name. The `6` at the beginning of the file tells me what stage of the inner loop this spec is for.

- Partially automated in ad-hoc scripts per language / repo.
- Inconsistent between management-plane and data-plane SDKs (especially version bump timing and changelog handling).
- Error-prone when performed manually (missed version increments, stale README examples, inconsistent metadata presented across different files, like `pom.xml`, `_metadata.json`).
- A barrier for integrating an AI agent workflow because required parameters and step order are not encapsulated as composable tools.

### Why This Matters

Without a unified tool layer:

- Contributors spend time re-learning differing per-language conventions.
- CI or release pipelines fail late due to missing metadata or mismatched versions.
- Agents cannot reliably assist because state transitions ("now bump version", "now update changelog") aren't formalized.
- Management-plane vs data-plane discrepancies introduce confusion (e.g., when to bump versions or generate changelog entries).

> Formalizing Stage 6 as four cohesive tools improves reliability, supports automation-first workflows, and prepares for future LLM-assisted changelog generation.

Reviewer 7: Keeping it generic
```suggestion
Formalizing Stage 6 as cohesive tools improves reliability, supports automation-first workflows, and prepares for future LLM-assisted tasks
```

---

## Design Proposal

### Overview

Implement four individual CLI + MCP tools. Each tool:

1. Resolves package path & language.
2. Loads the configuration for the language.
3. Chooses inline command over script path if both present.
4. Executes the relevant sub-script(s) in an ordered, fault-isolated manner.
5. Aggregates structured result JSON with a `result`, human-readable `message`, and optional `next_steps` hint guiding subsequent tool invocation.

Provides a singular CLI + MCP command that orchestrates the four tools (update-ci → update-changelog → update-version → update-metadata) and returns a combined JSON summary.

### Detailed Design

The four tools are intentionally small, composable units. Each emits a machine-readable JSON payload to enable deterministic chaining by an agent or CI workflow. All tools share common behaviors:

| Area | Expected Behavior |
|--------|-------------------|
| Invocation Mode | CLI and MCP (agent) alias; same semantics |
| Input Path Validation | Must exist and contain a recognizable SDK language structure; errors early if not |

> | Config Resolution | Load `swagger_to_sdk_config.json`; prefer `inline command` over `script path` when both are configured |

Author: This is a shared configuration used by both the SDK validation check and the generation pipeline. We’ve considered renaming it, but that would require a migration plan for the consuming code.

> | Timeout (default) | 5 minutes per tool invocation |

Reviewer 7: I REALLY hope that none of the 4 scenarios you listed takes more than seconds. What is the reason for 5min?

| Output Schema | JSON with fields: result, message (see schema below) |
| Idempotency | Multiple successive runs produce identical filesystem state except for timestamps/logs |

> | Script Implementation | MUST be PowerShell `.ps1`. SHOULD accept named parameters (e.g., `sdkRepoPath`). MUST use exit code `0` for success/noop; non-zero only on outright failure. Legacy non-PS scripts should be wrapped by a small PowerShell shim. Scripts MUST write a concise stderr line prefixed with `ERROR:`; include suggested mitigation steps and links to docs in the message where possible. |

Author: That's correct. Script is only invoked when the common case doesn't apply.

| Script Execution Environment | All scripts are executed using PowerShell Core (`pwsh`) terminal |
| Next Steps Hint | Plain language phrase embedded in `message` (e.g., `next_steps: update version`) |
| Missing Required Tools | If a required external tool is not present (for example: `dotnet`, `gofmt`, `pwsh`), prompt the user to run the `verify-setup` tool which validates and documents missing prerequisites. |

#### 1. Changelog Update Tool

> Name (CLI): `azsdk package update changelog`

Reviewer 8: IMO I think this should be "update changelog-content" or something and all it does is update the content

Author: Agree, I thought about using it since the update only involves the content.

Author: Updated.

> Name (CLI): `azsdk package update-changelog-content`

Reviewer: I had suggestion for name update but can be discussed on #12564.

Name (MCP): `azsdk_package_update_changelog`

> Purpose: Ensure changelog is auto-generated and has an entry for the upcoming release (management-plane) or provide guidance (data-plane) where manual editing is expected. This tool only modifies changelog content and does not change release date or version numbers.

Reviewer 7: this is the ultimate goal, but it would be better if the spec right now reflects what needs to be implemented for `Scenario 1` in which case, the purpose would be different, but each language will be able to know what the expectation is of what they need to implement

Reviewer 9: From mgmt side work on update changelog, We already have update changelog logics implemented in current generation script in CI. (except dotnet /cc @haiyuazhang). So our work is to split existing scripts, extract the logics, write it using powershell, and test. 

The API on how language script integrates with MCP can be found here: https://github.com/Azure/azure-sdk-tools/blob/f152f2e8df4208c3ab7fc211b4620d891365a748/tools/azsdk-cli/docs/specs/6-package-updater.spec.md#update-changelog-script

e.g. I'm working on the PR to update changelog script. https://github.com/Azure/azure-sdk-for-java/pull/47079/files

Author: > this is the ultimate goal

This is the goal of Scenario 1. For mgmt-plane SDKs, auto-generation is the goal; while for data-plane, the goal is to provide guidance.

Inputs:

- `--package-path <abs path>` (required)

Execution Steps (Mgmt-Plane):

1. Invoke changelog generation script or tool.

> 2. Insert (or update) latest section with release date placeholder / actual date.

Author: Here what I describe is more about the current behavior of the changelog generation tool mgmt-plane SDK uses. 
I agree we should have only one place to set the release date and version. That's the `update-version` tool.

I'll call out this behavior should be deleted from current changelog generation tool.

Execution Steps (Data-Plane):

> 1. Return `noop` result advising manual edit and future version bump timing.

Author: Do we already have a guideline doc?

> 1. Return `noop` result advising manual edit and future version bump timing.

Reviewer: I believe every language would at least have formats with content headers and specifications that need to be in the Readme, we can/should link to that.

Outputs:

```json
{"result":"succeeded", "message":"Changelog entry added for 2025-10-16", "next_steps": "update version"}
```

```json

> {"result":"noop", "message":"Data-plane changelog untouched", "next_steps": "Manually edit the changelog, and update version"}

Author: I'd prefer to keep it since it's expected to be consumed by the agent. I'll add the link when we have the doc in place.

{"result":"noop", "message":"Data-plane changelog untouched; manual edits required", "next_steps": "update version"}
```

Failure Modes:

> - Missing `CHANGELOG.md` → create template (mgmt-plane) or noop (data-plane).

Reviewer 10: I wonder if we should consider recovery here rather than continuing down the "yell at someone and make them fix it" when we have a problem.    Here, we could attempt to grab the latest changelog from the latest release tag and if we don't have a release, just create it.

Not saying that it needs to be this phase, but I think we need to shift our mindset about how we design tools.   They should be self-healing and find the pit of success when possible.   The approach of failing and making the caller's problem to fix is a big reason why our end-to-end experience has friction and why our team gets pulled into support.

//cc: @ronniegeraghty, @maririos for additional thoughts.

Author: Added the feedback to the spec.

- Script failure → `failed` with error output.

##### Update Changelog script

When the tool executes the language-specific `update-changelog` script it will pass the following parameters. All paths are absolute.

| Parameter | Description | Type | Required |
|-----------|-------------|-----:|:--------:|
| `SdkRepoPath` | Absolute path to the root folder of the local SDK repository. | string | Yes |
| `PackagePath` | Absolute path to the root folder of the local SDK project (package). | string | Yes |

Note: The script only needs to cover mgmt-plane.

**Example invocations:**

```powershell
update-changelog.ps1 \
      -SdkRepoPath C:\dev\repos\azure-sdk-for-net \
      -PackagePath C:\dev\repos\azure-sdk-for-net\services\newservice
```

#### 2. Version Update Tool

> Name (CLI): `azsdk package update version`

Author: In milestone 1, for data-plane SDK scenario, both 'update-changelog' and 'update-version' tools return `noop`, so they won't actually modify the changelog or version data.

Name (MCP): `azsdk_package_update_version`

Purpose: Increment and propagate package version across all authoritative version surfaces (e.g., `pom.xml`, `_version.py`, `*.csproj`, `package.json`, `go.mod`, `CHANGELOG.md`).

Inputs:

> - `--package-path <abs path>` (required)

Reviewer 10: How do we infer the repo path?

> - `--package-path <abs path>` (required)

Reviewer 7: why is this not true for the changelog tool?

> - `--package-path <abs path>` (required)

Reviewer 10: It makes sense.   It does mean that we've got a hard assumption that any package we're operating on is in the repository with the tooling.   You could not, for example, use the DevEx tool against the Microsoft MCP package, as that would cross repositories.

> Purpose: Increment and propagate package version across all authoritative version surfaces (e.g., `pom.xml`, `_version.py`, `*.csproj`, `package.json`, `go.mod`, `CHANGELOG.md`).

Author: There're more, but the purpose here is just to list some examples. Every language owner has better understanding of which files will be included. We could ask language teams to call them out explicitly if necessary.

- `--release-type <stable|beta>` (optional)
- `--version` (optional)

> - `--release-date` (optional)

Author: I'll call out both the release date and version should only be updated by this tool.

Execution Steps:

1. Compute a version based on the given information (mgmt-plane).

> 2. Apply updates across language-specific files.

Author: Sounds good, and separated them in two sections.

3. Validate consistency (all files now share identical new version).
4. For data-plane outside release: return `noop`.

> 5. For data-plane prepare release: use current version from the metadata file or from user input (support may be post-milestone-1).

Reviewer 10: Same question that I had.

> 5. For data-plane prepare release: use current version from the metadata file or from user input (support may be post-milestone-1).

Author: The tool itself doesn’t track context about the user’s intent. That will be determined by leveraging Copilot instructions, and once the user confirms it's for release purpose, this tool will be invoked.

> 5. For data-plane prepare release: use current version from the metadata file or from user input (support may be post-milestone-1).

Author: You’re right. The tool always sets the version whenever it’s invoked. What I meant is that, in the begining of local generation, the agent should confirm the user’s intent. After generation when run `package update`, the agent should call the `update-version` tool only if the purpose is a release.

> > **Note:** In milestone 1, Java management-plane packages will follow the same version update strategy as data-plane packages. Adjustments for Java-specific workflows may be introduced in later versions.

Reviewer 7: this should be included in the exclusion section

Author: I included this in the exception too.

```

```json
{"result":"noop", "version": "", "release-date": "", "message":"no version bump performed", "next_steps": "update metadata"}
```

Failure Modes:

- Version parse error (non-semver) → fail with offending token.
- Partial update (one file locked) → `partial` list and errors.

Idempotency: Re-running after success yields `succeeded` with `no changes` message when already at target version.

##### Update Version script

When the tool executes the language-specific `update-version` script it will pass the following parameters. All paths are absolute.

| Parameter | Description | Type | Required |
|-----------|-------------|------:|:--------:|
| `SdkRepoPath` | Absolute path to the root folder of the local SDK repository. | string | Yes |
| `PackagePath` | Absolute path to the root folder of the local SDK project (package). | string | Yes |
| `SdkReleaseType` | Specifies whether the next version is `beta` or `stable`. | string | No |
| `SdkVersion` | Specifies the next version to set (explicit version string). | string | No |

> | `SdkReleaseDate` | Release date to write into changelog (YYYY-MM-DD). | string | No |

Reviewer 7: brainstorming here.... should this be it's own cli? `setReleaseDate` :|
Because with this option here we are saying that if they update the version, then they add a release date, and maybe that wasn't their intention.
If this is only for management plane where this won't be an issue then you can ignore this.
But if this is setting expectation on what data plane will do in the future, then we should talk

Note: For management-plane (mgmt) packages, at least one of `SdkVersion` or `SdkReleaseType` MUST be provided; the script will fail with an error if neither is supplied or validation fails.

**Example invocations:**

```powershell

> update-version.ps1 \

Reviewer 8: For data plane is this doing anything more than wrapping the prepare release script? Does it do more in mgmt plane? It seems unnecessary to define a new script

Author: This script is for mgmt-plane only. For data-plane, I'll check whether we can call the `prepare-release` script directly or any adjustments are needed after m1.

> Note: For management-plane (mgmt) packages, at least one of `SdkVersion` or `SdkReleaseType` MUST be provided; the script will fail with an error if neither is supplied or validation fails.

Reviewer 9: I agree with @maririos, `sdkReleaseType` should always be default to `beta`. And I believe engsys could just pass the `sdkReleaseType` (which defaults to `beta`) to language script, so languages do not need to have the duplicate logic to handle default release type.

      -SdkRepoPath C:\dev\repos\azure-sdk-for-net \
      -PackagePath C:\dev\repos\azure-sdk-for-net\services\newservice \
      -SdkReleaseType stable \
      -SdkReleaseDate 2025-10-17
```

#### 3. Metadata Update Tool

Name (CLI): `azsdk package update-metadata`

Name (MCP): `azsdk_package_update_metadata`

Purpose: Update additional language-specific metadata and documentation, such as `_metadata.json`, `pom.xml` etc., export API surface (.NET).

Inputs:

- `--package-path <abs path>` (required)

Execution Steps:

1. Apply updates to authoritative metadata files (`_metadata.json`, `pom.xml`, `package.json`, `go.mod`). Only overwrite a file when the new content differs from the existing content. If an expected file is missing, create a minimal safe skeleton when possible and warn otherwise.

> 2. Return a structured summary in the tool JSON result including `changed_files` (array of relative paths) and `changed_files_count` (integer). When no files were modified return `changed_files: []` and `changed_files_count: 0` and set `message` to indicate "no changes" (`changed_files` may be supported post-milestone-1).

Author: Please see [this thread](https://github.com/Azure/azure-sdk-tools/pull/12563#discussion_r2453047905) for the discussion with Jesse. We may revist this while planning for next milestone.

Outputs:

```json

> {"result":"succeeded", "changed_files":["..\\pom.xml","C:\\dev\\repos\\azure-sdk-for-java\\pom.xml"], "changed_count": 2, "message":"All relevant pom.xml files have been updated."}

Author: The initial thought is to return a small, machine-readable summary that the agent/CI can consume without invoking git or scanning the repo. This isn't planned for M1, and we can revisit or remove this feature when planning for M2.

```

```json
{"result":"failed","message":"API export failed: missing dotnet tool", "next_steps": "call verify-setup tool"}
```

Failure Modes:

- Missing expected file (e.g., `_metadata.json`) → report error.

##### Update Metadata script

When the tool executes the language-specific `update-metadata` script it will pass the following parameters. All paths are absolute.

| Parameter | Description | Type | Required |
|-----------|-------------|-----:|:--------:|
| `SdkRepoPath` | Absolute path to the root folder of the local SDK repository. | string | Yes |
| `PackagePath` | Absolute path to the root folder of the local SDK project (package). | string | Yes |

**Example invocations:**

```powershell

> update-metadata.ps1 \

Reviewer 8: Curious- what's driving the motivation behind the consistent integration point being the script? In the other tools that have implementations in the CLI there are classes for each language and that's where the integration is - https://github.com/Azure/azure-sdk-tools/blob/main/tools/azsdk-cli/Azure.Sdk.Tools.Cli/Services/Languages/DotNetLanguageSpecificChecks.cs, I feel like it should be the same here. Then for things like normalizing paths and finding repo root, there are already helpers for those in the cli

> update-metadata.ps1 \

Reviewer 10: I'm a +1 to Maddy's thought.   We should have a single common approach for invoking language-specific paths.   If there is prior art using classes, we should follow that pattern everywhere.   There's no reason why a class-based solution couldn't invoke the pipeline scripts directly.    In my opinion better that then having a wrapper script invoke an existing script and diverging the approach.

//cc: @praveenkuttappan, @benbp

> update-metadata.ps1 \

Reviewer 11: > We should have a single common approach

We do and we also have a doc on how you can use it: https://github.com/Azure/azure-sdk-tools/blob/main/tools/azsdk-cli/docs/per-language.md. This tool can create its own interface for whatever language specific operations it has, just like the test runner, checks, and (soon) samples generation does. Personally I believe keeping things in the CLI app is much easier to reason about than having PowerShell scripts in each of the SDK repos.

> update-metadata.ps1 \

Author: > I may just be out of the loop but I'm not familiar with this, where is the framework for that?

@m-redding this is [the slide](https://microsoft-my.sharepoint.com/:p:/p/raychen/IQCk5tKXJhk_Qaz7Of_0hqbfARogaLG19LDy0jTBAeOYypw?e=kegwuN) introducing the framework. Basically, it has a configuration in every language repo to integrate the language specific code.

> update-metadata.ps1 \

Reviewer 8: @weshaggard I see what you mean about things being fragile and about working in different branches. And I agree that no one will remember to update the tool if they update the original script. But I also question if people will remember to update the script that calls the script even if it is in the same repo (unless there are CI checks for this?)

The other consideration is that some of the normalization can't be done nearly as easily in PowerShell. I feel like I would probably want to do some of that normalization in the tool anyway so I can take advantage of shared helpers and BCL types. And then the script that calls the script would be less helpful.

That being said if this is the way we want to move forward, then that's fine, but IMO we should probably be consistent across tools.

> update-metadata.ps1 \

Reviewer 7: This is very interesting. Would it be wrong to say that as we move into a world where we want automation to be more present and the generation of the SDKs to be simplified, to offer a centralized approach on the core of the team?
At the same time, we have evidence where the script approach works. 
It makes it harder for people to trace errors which is what I hear Jesse is pointing out?

> update-metadata.ps1 \

Reviewer 10: > I think we should have a meeting to talk through the details to figure out how to balance all the trade-offs appropriately.

Agreed.   Lots of back-n-forth are needed, and this is one long comment.  😄 

> The reason I like the wrapper script is that it lives with the source and is much easier to version and update as needed, BTW I'm not hung up on it being PS just that it lives closer to what it is coupled to.

I agree with the goal, but I'm viewing the locality in the inverse.   I would prefer to see things that exist just for the DevEx platform be local to the platform.   I also don't care about the form, but I really don't want to see the language repo continue to accrue wrappers for platform X, Y, and Z.    

If we're going to do this, then I'd really like to advocate that we update the base script so that we can call it locally, in the pipelines, and via the DevEx platform directly rather than adding more levels of opaque indirection to wrap.   The flow through the repo for tools like this is already dense and very often unclear.   We spend days digging through the abstractions sometimes when troubleshooting.

> update-metadata.ps1 \

Reviewer 2: I would also love to see all these scripts become the new interface for interacting with the repo. That means they can be run locally, in our pipelines, in the Agent MCP tools, etc. That is definitely something I would like to see no matter where we land on the implementation.

> update-metadata.ps1 \

Reviewer 9: >  I really hate the alternative pattern of invoking foo.ps1 that does nothing but normalize parameters and invoke bar.ps1. The number of scripts and the opaqueness of how each gets called by different platforms is already difficult.

@jsquire I shared the similar concerns with you and raised it a few months ago at the begining of inner loop design into design group about defining a configuration and let each language integrate using scripts. Copy some of my comments here: 
_The design to me is not a very good 'agentic' design, it is more of a traditional design to define clear configuration/interface and integrate together. But in the 'open agentic web' world, everything is tool, and the agent will schedule the correct tool to complete the work, rather than we need to define a configuration/script and then integrate together._ 

More details here: https://teams.microsoft.com/l/message/19:meeting_YmI2NTNjZjctMjNhOC00MDViLWFlY2QtYzVhNjNkYzhjNGI3@thread.v2/1752806634239?context=%7B%22contextType%22%3A%22chat%22%7D

But I was told by @raych1 that the design was decided at that time. So I volunteered to work with @raych1 on the build sdk tasks, which use scripts to integrate to prove the concept, because Shanghai side language owners wrote the existing scripts (except dotnet probably),  e.g. Java ones: https://github.com/Azure/azure-sdk-for-java/blob/main/eng/automation. Though I had concerns with the approach. 😄  /cc @lmazuel @maririos @ArthurMa1978 @lirenhe 

So I really like to have furthur discussion on this script approach, and to get agreement and alignment! 😄 

---------------

@maririos @lmazuel @raych1 @weshaggard, **I saw the design seems not approved by all, so for all 'package update' tasks assigned to Shanghai(refer category 6 here: https://github.com/orgs/Azure/projects/865/views/17), may I ask shall we hold on the implementation of these scripts until the design is agreed/finalized?**(/cc @lirenhe @ArthurMa1978  @msyyc @MaryGao @tadelesh @haiyuazhang) 

**More background:** I was working with @raych1 on the design to provide feedbacks from mgmt side as Shanghai language owners wrote previous scripts and I was told the design is reviewed by Redmond. Shanghai side agreed to implement the scripts because our language owners wrote previous scripts and may have more context (except dotnet probably). @raych1 requested us to do status update on these scripts implementation at Oct 15th mentioning the timeline is Oct 26th, which is quite tight thus Shanghai side is already prioritizing the tasks and is in progress of the scripts.  But **I am surprised to see the design is still not agreed by all.**  😢 

<img width="1483" height="345" alt="image" src="https://github.com/user-attachments/assets/ac243e16-f0dd-4fb3-9762-384f70502c3f" />

> update-metadata.ps1 \

Reviewer 9: > Thanks for all the detailed feedback this is a great conversation and like everything is a matter of balancing trade-offs. I will try to respond to some of the comments here, but I think we should have a meeting to talk through the details to figure out how to balance all the trade-offs appropriately. @raych1 would you mind setting up something?

@weshaggard @raych1, as Shanghai side is working on the implementation of the script. Is it possible to leverage the Friday's(your Thursday afternoon) inner loop sync up meeting to discuss, so that we can be on the same page? Thanks a lot!
/cc @maririos @lmazuel

> update-metadata.ps1 \

Author: @m-redding @timovv @haolingdong-msft 
As we agreed on the Ben's proposal, the `update-metadata` tool will invoke the `RepositoryService` to run a powershell script or fall back to the built-in C# code in CLI. Let me know if there're any gaps.

> update-metadata.ps1 \

Reviewer 11: My understanding was that we agreed to use the RepositoryService pattern for calling _existing_ PowerShell scripts. What's being proposed here, if I understand correctly, is a _new_ PowerShell script and this logic should instead be implemented in the CLI in C#.

> update-metadata.ps1 \

Reviewer 8: Ben's proposal specifically states that no contract is provided for parameters, so I don't think we should include parameters in the contract.

> update-metadata.ps1 \

Reviewer 8: With what you have here it's not possible for me to call export api without wrapping it because this contract doesn't match the parameters. I need the service directory and not the path. https://github.com/Azure/azure-sdk-for-net/blob/main/eng/scripts/Export-API.ps1#L5-L7

This is the current implementation that the Shanghai team is working on https://github.com/Azure/azure-sdk-for-net/pull/53458/files which requires brittle string parsing

##### Update CI script

When the tool executes the language-specific `update-ci` script it will pass the following parameters. All paths are absolute.

| Parameter | Description | Type | Required |
|-----------|-------------|-----:|:--------:|
| `SdkRepoPath` | Absolute path to the root folder of the local SDK repository. | string | Yes |
| `PackagePath` | Absolute path to the root folder of the local SDK project (package). | string | Yes |

**Example invocations:**

```powershell
update-ci.ps1 \
      -SdkRepoPath C:\dev\repos\azure-sdk-for-net \
      -PackagePath C:\dev\repos\azure-sdk-for-net\services\newservice
```

#### 5. Aggregated Tool (chained execution)

Name (CLI): `azsdk package update all`

Name (MCP): `azsdk_package_update_all`

> **Note:** Support for this tool is planned for post‑milestone-1.

Purpose: provide a single entrypoint that orchestrates the four tools in a deterministic sequence for common release-prep workflows.

Sequence: update-ci -> update-changelog -> update-version -> update-metadata

Execution Steps:

- Validate the provided `packagePath` and fail fast if invalid.
- Run `update-ci` first to ensure CI config is present/updated.

> - Run `update-changelog` next; if it returns `failed`, stop and return the error.

Reviewer 10: I think we need more focus on "how can we automatically apply fixes" rather than just "its the caller's problem"

      C -->|Yes| E[Discover SDK repository root] 
    
      E --> F{Repository root found?}
      F -->|No| G[Return failure: Failed to discover SDK repo]
      F -->|Yes| H[Get update-changelog-script configuration] 
    
      H --> I{Configuration retrieval successful?}
      I -->|No| J[Return failure: Failed to get update-changelog-script configuration] 
      I -->|Yes| K{Configuration type?}
    
      K -->|Command| L[Prepare update-changelog command] 
      K -->|ScriptPath| M[Prepare update-changelog script] 
    
      L --> N[Execute update-changelog command/script] 
      M --> N
    
      N --> O{update-changelog process completed?} 
      O -->|Failed| P[Return failure with exit code and output]
      O -->|Success| Q[Return success]
    
      style A fill:#e1f5fe
      style Q fill:#c8e6c9
      style D fill:#ffcdd2
      style G fill:#ffcdd2
      style J fill:#ffcdd2
      style P fill:#ffcdd2
```

> **Note:** Other tools (version update, metadata update, CI update) follow a similar flowchart pattern.

---

## Open Questions

---

## Success Criteria

This tool set is complete when:

- [ ] All four tools emit structured JSON with consistent schema.

> - [ ] Management-plane package run updates version + changelog automatically.

Reviewer 7: but here we would need to list the limitations or exceptions too.
might be better to either call out specifically every language per category or summarize that all languages and categories that are not in the exceptions should work.

Author: Added exceptions.

- [ ] Data-plane run returns advisory for changelog and version.
- [ ] Cross-language scripts execute without manual intervention.
- [ ] Errors surface actionable mitigation text.

---

## Agent Prompts

### Scenario 1: Update Management-Plane Package (existing SDK) for Release

**Prompt:**

```text
Update the package at /home/dev/sdk/healthdataaiservices/Azure.ResourceManager.HealthDataAIServices for a stable release.
```

**Expected Agent Activity:**

1. Run changelog update (auto-generation script).
2. Run version update with inferred `--release-type stable`.
3. Run metadata update (README, API export, formatting if needed) and summarize JSON results.

> ### Scenario 2: Update Version for Stable Release (Data-Plane)

Author: The version will be updated every time the tool is invoked. As mentioned above, the user intent will be determined by instructions instead of in the tool itself.

> ### Scenario 2: Update Version for Stable Release (Data-Plane)

Author: When users start the workflow of SDK generation, the agent will prompt them like:
"Is this generation for a release? If yes, the tool may update release date and bump versions for the created package. Proceed as release? (yes/no)". If the user answers "yes" the agent runs the `update-version` tool; otherwise, it will not be invoked.

**Prompt:**

```text

> Update the version for package at /repo/sdk/healthdataaiservices/Azure.Health.Deidentification.

Reviewer 10: This should also set the release date in the change log.   For release purposes, that should be considered part of the release version.

Author: That's true.  The tool will prompt user to confirm on the date even if the user prompt doesn't mention the release date.

```

**Expected Agent Activity:**

1. Confirm with user on the version to use.

> 2. Run version update with the confirmed version.

Reviewer 8: I think above it says version update noops unless it's a release, how will it know that this case it should update the version? Or will the tool delegate to the prepare release script if a version and date are passed in?

## Table of Contents

- [Definitions](#definitions)
- [Background / Problem Statement](#background--problem-statement)
- [Goals and Exceptions/Limitations](#goals-and-exceptionslimitations)
- [Design Proposal](#design-proposal)
- [Open Questions](#open-questions)
- [Success Criteria](#success-criteria)
- [Agent Prompts](#agent-prompts)
- [CLI Commands](#cli-commands)
- [Implementation Plan](#implementation-plan)
- [Testing Strategy](#testing-strategy)
- [Documentation Updates](#documentation-updates)
- [Metrics/Telemetry](#metricstelemetry)

---

## Definitions

- **Stage 6**: The standardized SDK preparation stage after code/tests/samples validation and before creating the release preparation PR.
- **Metadata and Documentation Update**: Consolidated set of actions to ensure non-source assets (README, changelog, CI config, version numbers, additional descriptors) are accurate and consistent.
- **Tool**: A discrete CLI (and MCP agent-exposed) command implementing one cohesive post-generation preparation task.
- **Inline Script Command**: A single shell command provided directly in config that overrides a script path.

---

## Background / Problem Statement

### Current State

> After SDK code generation, build, tests, and sample validation, engineers must perform repetitive cross-language housekeeping: updating changelogs, synchronizing README and language-specific files, adjusting package versions, exporting API surfaces (.NET), and inserting or validating CI configuration. Today these tasks are:

Reviewer 10: ```suggestion
After code generation, build, tests, and sample validation, engineers must perform repetitive cross-language housekeeping: updating changelogs, synchronizing README and language-specific files, adjusting package versions, exporting API surfaces (.NET), and inserting or validating CI configuration. Today these tasks are:
```

nit:  Azure SDK is the all-up brand.  Each individual part is a "library" or "package" not an SDK.

> - Partially automated in ad-hoc scripts per language / repo.

Author: That's true. These are part of the problem statement. The specific tasks to be done are described in every tool's design section.

- Inconsistent between management-plane and data-plane SDKs (especially version bump timing and changelog handling).
- Error-prone when performed manually (missed version increments, stale README examples, inconsistent metadata presented across different files, like `pom.xml`, `_metadata.json`).
- A barrier for integrating an AI agent workflow because required parameters and step order are not encapsulated as composable tools.

### Why This Matters

Without a unified tool layer:

- Contributors spend time re-learning differing per-language conventions.
- CI or release pipelines fail late due to missing metadata or mismatched versions.
- Agents cannot reliably assist because state transitions ("now bump version", "now update changelog") aren't formalized.

> - Management-plane vs data-plane discrepancies introduce confusion (e.g., when to bump versions or generate changelog entries).

Reviewer 12: Could you share more about the difference on how data plane SDK do version bump and generate changelog?  is this done manually or using a another tool?

Formalizing Stage 6 as four cohesive tools improves reliability, supports automation-first workflows, and prepares for future LLM-assisted changelog generation.

---

## Goals and Exceptions/Limitations

### Goals

- [ ] Provide four composable, idempotent tools exposed via CLI and MCP agent: CI Update, Changelog Update, Version Update, Metadata Update.
- [ ] Normalize execution order and expected outputs across all supported languages.
- [ ] Encode management-plane vs data-plane behavioral differences explicitly (especially version & changelog logic).
- [ ] Emit structured JSON results (status + message + next steps hints) enabling orchestration by AI agents and pipelines.
- [ ] Support configuration-driven per-language scripting (path or inline) without recompiling the tool.

### Exceptions and Limitations

> #### Exception 1: Data-Plane Deferred Version Bump

Reviewer 12: I think our goal is to make data plane and mgmt. plane work to be consistent.
So curious is there any specific reason that those work could not be done earlier in the process?

cc @lmazuel

Reviewer 10: @lirenhe:  Data plane packages follow a very different development lifecycle than management packages.   They generally do not release after a single pull request, and changes accumulate in `main` over a period of time.    The release date and version are not decided upfront - they're decided on after architect review and feedback.     The choice of version is an outcome of architect discussion and is nuanced; not something that can be calculated algorithmically.

There's no alignment in the process and approach between management and data plane packages.   Thus, we do not have the opportunity to force consistency in the tooling.

**Description:** Data-plane packages typically defer version adjustment and release date update until the release preparedness phase, not immediately after Stage 6 baseline tasks.
**Impact:** The `version update` tool may be a no-op for data-plane flows outside release prep.
**Workaround:** Initial version of milestone 1 remains user decided and manual edits.

#### Exception 2: Changelog Auto-Generation (Data-Plane)

**Description:** No automatic generation for data-plane; manual editing expected.
**Impact:** Tool returns success with a `next_steps` message prompting manual edits.
**Workaround:** Future LLM integration (post milestone 1).

#### Language-Specific Limitations

| Language   | Limitation | Impact | Workaround |
|------------|------------|--------|------------|

> | JavaScript | Code formatting isn't hooked up to the emitter | Noisy diffs or linter failures | Run code formatting step in the `update-metadata.ps1` script |

Reviewer 12: @MaryGao,   could you help to confirm?

1. Create the corresponding YAML file, like `ci.yml`, which will be used to setup pipeline.
2. Update the relevant CI YAML file accordingly
3. Return the changed file name with the result.

Outputs (examples):

```json
{"result":"succeeded", "ci-file": "../ci.yml", "message":"CI files created: ci.yml", "next_steps": "update changelog"}
```

```json
{"result":"failed", "message":"script exited 1: missing pwsh"}
```

Failure Modes:

- Create file fails due to insufficient permissions → return `failed` with error.

##### Update CI script

When the tool executes the language-specific `update-ci` script it will pass the following parameters. All paths are absolute.

| Parameter | Description | Type | Required |
|-----------|-------------|-----:|:--------:|
| `SdkRepoPath` | Absolute path to the root folder of the local SDK repository. | string | Yes |
| `PackagePath` | Absolute path to the root folder of the local SDK project (package). | string | Yes |

**Example invocations:**

```powershell
update-ci.ps1 \
      -SdkRepoPath C:\dev\repos\azure-sdk-for-net \
      -PackagePath C:\dev\repos\azure-sdk-for-net\services\newservice
```

#### 5. Aggregated Tool (chained execution)

> Name (CLI): `azsdk package update-all`

Reviewer 9: May I know the reason for having both separate tool and aggreated tool?

Author: Aggregated tool is expected to integrate to automation pipeline scenario and release scenario.
